pkg/component/kubernetes/apiserver/authentication.go

```diff
@@ -0,0 +1,219 @@
+// SPDX-FileCopyrightText: 2024 SAP SE or an SAP affiliate company and Gardener contributors
+//
+// SPDX-License-Identifier: Apache-2.0
+
+package apiserver
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"strings"
+
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	genericapiserverv1alpha1 "k8s.io/apiserver/pkg/apis/apiserver/v1alpha1"
+	"k8s.io/utils/ptr"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/yaml"
+
+	gardencorev1beta1 "github.com/gardener/gardener/pkg/apis/core/v1beta1"
+	kubernetesutils "github.com/gardener/gardener/pkg/utils/kubernetes"
+	versionutils "github.com/gardener/gardener/pkg/utils/version"
+)
+
+const (
+	volumeNameStructuredAuthenticationConfig = "authentication-config"
+
+	volumeMountStructuredAuthenticationConfig = "/etc/kubernetes/structured/authentication"
+	volumeMountPathOIDCCABundle               = "/srv/kubernetes/oidc"
+
+	configMapAuthenticationConfigDataKey = "config.yaml"
+)
+
+// reconcileConfigMapAuthenticationConfig reconciles the ConfigMap containing the authentication configuration.
+func (k *kubeAPIServer) reconcileConfigMapAuthenticationConfig(ctx context.Context, configMap *corev1.ConfigMap) error {
+	if versionutils.ConstraintK8sLess130.Check(k.values.Version) && k.values.AuthenticationConfiguration != nil {
+		return errors.New("structured authentication is not available for versions < v1.30")
+	}
+	if k.values.AuthenticationConfiguration != nil && k.values.OIDC != nil {
+		return errors.New("oidc configuration is incompatible with structured authentication")
+	}
+	if value, ok := k.values.FeatureGates["StructuredAuthenticationConfiguration"]; (ok && !value) ||
+		(k.values.AuthenticationConfiguration == nil && k.values.OIDC == nil) ||
+		versionutils.ConstraintK8sLess130.Check(k.values.Version) {
+		return nil
+	}
+
+	var authenticationConfig string
+	if k.values.AuthenticationConfiguration != nil {
+		authenticationConfig = *k.values.AuthenticationConfiguration
+	}
+	if k.values.OIDC != nil {
+		oidcAuthenticationConfig, err := computeAuthenticationConfigRawConfig(k.values.OIDC)
+		if err != nil {
+			return err
+		}
+
+		authenticationConfig = oidcAuthenticationConfig
+	}
+
+	configMap.Data = map[string]string{configMapAuthenticationConfigDataKey: authenticationConfig}
+	utilruntime.Must(kubernetesutils.MakeUnique(configMap))
+	return client.IgnoreAlreadyExists(k.client.Client().Create(ctx, configMap))
+}
+
+func computeAuthenticationConfigRawConfig(OIDC *gardencorev1beta1.OIDCConfig) (string, error) {
```

Comment:
From my point of view this is the function where things can go wrong, i.e. when converting one config to the other. Can we make this or another function exported returning `&AuthenticationConfiguration` and unit testing it with all edge cases of the previous OIDC configs, the email claim, the "-" character, defaulting for configs, etc.

WDYT?

Reply:
Sure. I had already added the edge cases to the `pkg/component/kubernetes/apiserver/apiserver_test.go` unit test, but I do see a good reason to have them separate as well. I have now added them in my latests commits.